### ISSUE TITLE
Client credentials should be authenticated where possible

### ISSUE THREAD

**Issue by jordanbtucker** [daskeyboard/daskeyboard.io#11]
Client credentials (`client_id` and `client_secret`) can be used to obtain an access token, but beyond that, they aren't used for authentication. The purpose of OAuth client credentials is to authenticate third party services like IFTTT and Zapier, not to authenticate end users.

According to the [OAuth RFC](https://tools.ietf.org/html/rfc6749#section-3.2.1), client credentials are used for:

> - Enforcing the binding of refresh tokens and authorization codes to
>   the client they were issued to.  Client authentication is critical
>   when an authorization code is transmitted to the redirection
>   endpoint over an insecure channel or when the redirection URI has
>   not been registered in full.
> - Recovering from a compromised client by disabling the client or
>   changing its credentials, thus preventing an attacker from abusing
>   stolen refresh tokens.  Changing a single set of client
>   credentials is significantly faster than revoking an entire set of
>   refresh tokens.
> - Implementing authentication management best practices, which
>   require periodic credential rotation.  Rotation of an entire set
>   of refresh tokens can be challenging, while rotation of a single
>   set of client credentials is significantly easier.

Note that this only applies to [confidential clients](https://tools.ietf.org/html/rfc6749#section-2.1) like web applications that communicate with the Q API on the server backend over secure channels (i.e. not in a user's browser). They must be required to [register their redirect URIs](https://tools.ietf.org/html/rfc6749#section-3.1.2.2), and they should be required to use HTTPS. I'm not sure if you require IFTTT and Zapier to do that at this point.

[Public clients](https://tools.ietf.org/html/rfc6749#section-2.1), like mobile apps, desktop apps, or browser apps, that communicate to the Q API directly (not through a proxy web application as described earlier) should not be required to authenticate, nor should they even be given a `client_secret`, since they cannot securely store it. Also, they should only be allowed to use [loopback redirect URIs](https://tools.ietf.org/html/rfc8252#section-8.3).

End users have no need for client credentials, and giving a [set to them by default](https://q.daskeyboard.com/account) can lead to confusion on how to use the API, especially when the client credentials are tied to their personal account rather than an [app they are developing](https://github.com/DasKeyboard/q/issues/9). End users can just use their `username` and `password` to authenticate against the API rather than having two sets of credentials that do the same thing.

On second thought, the `password` grant type should probably be disabled, but that's best described in another issue.

In conclusion, if you're not going to authenticate client credentials, then there's no need to have `client_secret`s at all.
